custom_components/waste_collection_schedule/waste_collection_schedule/source/okc_gov.py

```diff
@@ -0,0 +1,74 @@
+import json
+import requests
+
+from datetime import datetime, timedelta
+from waste_collection_schedule import Collection
+
+TITLE = "City of Oklahoma City"
+DESCRIPTION = (
+    "Source for okc.gov services for City of Oklahoma City"
+)
+URL = "https://www.okc.gov"
+COUNTRY = "us"
+TEST_CASES = {
+    "Test_001": {"objectID": "1781151"},
+    "Test_002": {"objectID": "2002902"},
+    "Test_003": {"objectID": 1935340},
+}
+HEADERS = {
+"content-type": "text/json",
+"user-agent": "Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/111.0",
+}
+ICON_MAP = {
+    "TRASH": "mdi:trash-can",
+    "RECYCLE": "mdi:recycle",
+    "BULKY": "mdi:sofa",
+}
+
+
+class Source:
+    def __init__(self, objectID):
+        self._resourceID = str(objectID)
+
+    def fetch(self):
+
+        s = requests.Session()
+        r = s.get(
+            f"https://data.okc.gov/services/portal/api/data/records/Address%20Trash%20Services?recordID={self._resourceID}",
+            headers=HEADERS
+        )
+
+        # api response can be inconsistent, so catch cases were it doesn't return a collection schedule
+        try:
+            json_data = json.loads(r.text)
+        except:
+            print("JSON Decode Error - invalid response returned from data.okc.gov")
```

Comment:
I'm not sure if this is actually better then just raising an exception with this message.
It looks like the raising of an exception will prevent the source from updating. Your current approach will update the events with an empty list. But you probably want to keep the old one on an exception.  

https://github.com/mampfes/hacs_waste_collection_schedule/blob/bedf19841b72739c89f5ad86e471f606dafdd54f/custom_components/waste_collection_schedule/waste_collection_schedule/source_shell.py#L130-L152

But I'm not that deep in the code base. So @mampfes probably needs to answer this.

Reply:
If the previous entries remain when an error occurs then I can see the advantage of removing the try:except and let source_shell catch it. The test result then looks like this:

```bash
Testing source okc_gov ...
  found 6 entries for Test_001
    2023-04-11 : Trash [mdi:trash-can]
    2023-04-11 : Recycle [mdi:recycle]
    2023-04-25 : Recycle [mdi:recycle]
    2023-05-09 : Recycle [mdi:recycle]
    2023-05-01 : Bulky [mdi:sofa]
    2023-06-05 : Bulky [mdi:sofa]
  found 6 entries for Test_002
    2023-04-11 : Trash [mdi:trash-can]
    2023-04-11 : Recycle [mdi:recycle]
    2023-04-25 : Recycle [mdi:recycle]
    2023-05-09 : Recycle [mdi:recycle]
    2023-05-01 : Bulky [mdi:sofa]
    2023-06-05 : Bulky [mdi:sofa]
  Test_003 failed: Expecting value: line 1 column 1 (char 0)
```

I added the JSON Decode message so the test script  was a bit more informative than `Expecting value: line 1 column 1 (char 0)`, but if an end-user looked in the logs they'd see the full description.

```bash
ERROR (SyncWorker_4) [waste_collection_schedule.source_shell] fetch failed for source City of Oklahoma City:
Traceback (most recent call last):
File "/config/custom_components/waste_collection_schedule/waste_collection_schedule/source_shell.py", line 134, in fetch
entries = self._source.fetch()
File "/config/custom_components/waste_collection_schedule/waste_collection_schedule/source/okc_gov.py", line 41, in fetch
json_data = json.loads(r.text)
File "/usr/local/lib/python3.10/json/__init__.py", line 346, in loads
return _default_decoder.decode(s)
File "/usr/local/lib/python3.10/json/decoder.py", line 337, in decode
obj, end = self.raw_decode(s, idx=_w(s, 0).end())
File "/usr/local/lib/python3.10/json/decoder.py", line 355, in raw_decode
raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```
I don't mind which approach we take.

Reply:
if you don't like the error message you could do something like this:
```python
        try:
            json_data = json.loads(r.text)
        except Exception as e:
            raise Exception("invalid response returned from data.okc.gov") from e
```
will raise something like:
```python
Traceback (most recent call last):
File "/config/custom_components/waste_collection_schedule/waste_collection_schedule/source_shell.py", line 134, in fetch
entries = self._source.fetch()
File "/config/custom_components/waste_collection_schedule/waste_collection_schedule/source/okc_gov.py", line 41, in fetch
json_data = json.loads(r.text)
File "/usr/local/lib/python3.10/json/__init__.py", line 346, in loads
return _default_decoder.decode(s)
File "/usr/local/lib/python3.10/json/decoder.py", line 337, in decode
obj, end = self.raw_decode(s, idx=_w(s, 0).end())
File "/usr/local/lib/python3.10/json/decoder.py", line 355, in raw_decode
raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
File "/config/custom_components/waste_collection_schedule/waste_collection_schedule/source_shell.py", line 134, in fetch
entries = self._source.fetch()
File "/config/custom_components/waste_collection_schedule/waste_collection_schedule/source/okc_gov.py", line 41, in fetch
Exception: invalid response returned from data.okc.gov
```

```
Testing source okc_gov ...
  Test_001 failed: invalid response returned from data.okc.gov
  Test_002 failed: invalid response returned from data.okc.gov
  Test_003 failed: invalid response returned from data.okc.gov
```



It seems like they use Incapsula  banyule_vic_gov_au has the same issue (see main README.md) there it just raises the JSONDecodeError. You might want to add this source there as well.

Reply:
That looks a bit more elegant.

Agreed, Incapsula is the source of the error - the call returns html with Incapsula content when it fails to return json.
The script uses the API call detailed in the docs, so the inconsistency is frustrating. Maybe it'd be less of in issue for an end-user as they're likely to only make one query per day, rather than several is quick succession.